hw/drivers/sensors/bma253/src/bma253.c

```diff
@@ -2714,65 +2987,114 @@ bma253_set_fifo_cfg(const struct bma253 * bma253,
         break;
     }
 
+    data |= (0x03 << 2);
+
     return set_register((struct bma253 *)bma253, REG_ADDR_FIFO_CONFIG_1, data);
 }
 
-int
-bma253_get_fifo(const struct bma253 * bma253,
-                enum bma253_g_range g_range,
-                enum fifo_data fifo_data,
-                struct accel_data * accel_data)
+
+struct sensor_data_subscriber_info {
+    sensor_data_func_t read_func;
+    void *read_arg;
+};
+
+    int
+bma253_read_and_handle_fifo_data(
+        struct bma253           *bma253,
+        enum fifo_data          fifo_data,
+        struct sensor_data_subscriber_info *sdsi)
 {
-    float accel_scale;
-    uint8_t size, iter;
-    uint8_t data[AXIS_ALL << 1];
-    int rc;
+    int     rc;
+    uint32_t i;
 
-    switch (g_range) {
-    case BMA253_G_RANGE_2:
-        accel_scale = 0.00098;
-        break;
-    case BMA253_G_RANGE_4:
-        accel_scale = 0.00195;
-        break;
-    case BMA253_G_RANGE_8:
-        accel_scale = 0.00391;
-        break;
-    case BMA253_G_RANGE_16:
-        accel_scale = 0.00781;
-        break;
-    default:
-        return SYS_EINVAL;
+    float   accel_scale;
+
+    uint8_t *ff_buf = bma253->pdd.fifo_buf;
+    uint32_t size;
+    uint32_t frm_size;
+
+    uint8_t ff_frm_cnt = 0; //current fifo frame counter
+    bool    ff_or = 0;  // fifo over run bit
+
+    struct accel_data   accel_data[AXIS_ALL];
+    struct sensor_accel_data sad;
+
+    enum bma253_g_range g_range = bma253->cfg.g_range;
+
+    struct sensor *sensor = &bma253->sensor;
+
+    rc = get_accel_scale(bma253, g_range, &accel_scale);
+    if (0 != rc) {
+        return rc;
     }
 
+    //zg: to improve
     switch (fifo_data) {
     case FIFO_DATA_X_AND_Y_AND_Z:
-        size = AXIS_ALL << 1;
+        frm_size = AXIS_ALL << 1;
         break;
     case FIFO_DATA_X_ONLY:
     case FIFO_DATA_Y_ONLY:
     case FIFO_DATA_Z_ONLY:
-        size = 1 << 1;
+        frm_size = 1 << 1;
         break;
     default:
         return SYS_EINVAL;
     }
 
-    rc = get_registers((struct bma253 *)bma253, REG_ADDR_FIFO_DATA, data, size);
-    if (rc != 0) {
+    rc = bma253_get_fifo_status(bma253, &ff_or, &ff_frm_cnt);
+    if (0 == rc) {
+        if (!ff_or) {
```

Comment:
why us if (!ff_or) when if statement is empty, if (ff_or) would be shorter without empty statement